File: complementary_dna.go
package main

import (
	"fmt"
	"strings"
)

func DNAStrand(dna string) string {
	result := ""
	pairs := map[string]string{"A": "T", "T": "A", "C": "G", "G": "C"}

	for _, letter := range dna {
		result += pairs[string(letter)]
	}
	return result
}

func BestPractice(dna string) string {
	replacer := strings.NewReplacer("A", "T", "T", "A", "G", "C", "C", "G")
	return replacer.Replace(dna)
}

func main() {

	test := make(map[string]string, 3)
	test["AAAA"] = "TTTT"
	test["ATTGC"] = "TAACG"
	test["GTAT"] = "CATA"

	for args, expected := range test {
		fmt.Println(DNAStrand(args) == expected)
	}

	for args, expected := range test {
		fmt.Println(BestPractice(args) == expected)
	}

}

File: complementary_dna.py
def DNA_strand(dna: str) -> str:
    result: str = ""
    for letter in dna:
        if letter == 'A':
            result += 'T'
        elif letter == 'T':
            result += 'A'
        elif letter == 'C':
            result += 'G'
        elif letter == 'G':
            result += 'C'
    return result

def best_practice(dna):
    pairs = {'A':'T','T':'A','C':'G','G':'C'}
    return ''.join(pairs[letter] for letter in dna)


def main():
    test = {
        "AAAA":"TTTT",
        "ATTGC": "TAACG",
        "GTAT":"CATA",
    }

    for par, expected in test.items():
        got = DNA_strand(par)
        try:
            assert(got == expected)
        except AssertionError:
            print(got, expected)

    for par, expected in test.items():
        got = best_practice(par)
        try:
            assert(got == expected)
        except AssertionError:
            print(got, expected)
    print("Tested !")

if __name__ == "__main__":
    main()
